### PR TITLE
WIP on ConvertingValueComparer

### PR DESCRIPTION
**Pull request by roji**
Prototype for https://github.com/dotnet/efcore/pull/35326#discussion_r1891361149

### CODE REVIEW

File: src/EFCore/ChangeTracking/Internal/ValueComparerExtensions.cs [hunk 1]
```diff
@@ -25,4 +25,35 @@ public static class ValueComparerExtensions
...
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public static ValueComparer? ComposeConversion(this ValueComparer? valueComparer, Type targetClrType)
+    {
+        if (valueComparer is null || valueComparer.Type == targetClrType)
+        {
+            return valueComparer;
+        }
+
+        if (targetClrType.IsNullableValueType() && !valueComparer.Type.IsNullableValueType())
```
Comment (AndriySvyryd) [inline src/EFCore/ChangeTracking/Internal/ValueComparerExtensions.cs:42]: Don't you also need to check that `targetClrType.UnwrapNullableType() == valueComparer.Type`?

**Comment by AndriySvyryd** [comment]
I still believe that for 9.0.2 https://github.com/dotnet/efcore/pull/35360 is a less risky fix and considering that we need to keep the legacy code for quirk, I don't see any downside to it.

**Comment by roji** [comment]
OK @AndriySvyryd, am OK with that. @maumar @AndriySvyryd you can close this PR or keep it, reusing the code here for the 10 fix if we end up going that way.

**Comment by AndriySvyryd** [comment]
Superceded by #35326

**Pull request closed without merging**